src/goto-programs/goto_contractor.h

```diff
@@ -41,6 +41,210 @@ class vart
   void setIntervalChanged(bool intervalChanged);
   const symbol2tc &getSymbol() const;
 };
+
+class Contractor
+{
+  ibex::Ctc *outer;
+  ibex::Ctc *inner;
+  unsigned int location;
+
+public:
+  Contractor(ibex::Ctc *c, unsigned int loc)
+  {
+    outer = c;
+    location = loc;
+    inner = get_complement_contractor(c);
+  }
+  Contractor() = default;
+  void set_outer(ibex::Ctc *outer)
+  {
+    Contractor::outer = outer;
+  }
+  void set_inner(ibex::Ctc *inner)
+  {
+    Contractor::inner = inner;
+  }
+  void set_location(unsigned int location)
+  {
+    Contractor::location = location;
+  }
+  ibex::Ctc *get_outer() const
+  {
+    return outer;
+  }
+  ibex::Ctc *get_inner() const
+  {
+    return inner;
+  }
+  unsigned int get_location() const
+  {
+    return location;
+  }
+
+private:
+  ibex::Ctc *get_complement_contractor(ibex::Ctc *c)
+  {
+    ibex::Ctc *complement_contractor = nullptr;
+    auto type = boost::core::demangle(typeid(*c).name());
```

Comment:
Unfortunately, that's the only way I found to determine the type in ibex.

Reply:
I just had a look at the API and all these objects inherit from ctc, so you can use the dynamic_cast that you already implemented to convert them safety.

It will look something like: 
```
if(auto ctc_compo = dynamic_cast<ibex::CtcCompo *>(c)) {
      ibex::Array<ibex::Ctc> list_of_contractors;
      for(auto &it : ctc_compo->list)
        list_of_contractors.add(*get_complement_contractor(&it));
      return new ibex::CtcUnion(list_of_contractors);
} else if (auto ctc_union = dynamic_cast<ibex::CtcUnion *>(c)) {
      auto ctc_union = dynamic_cast<ibex::CtcUnion *>(c);
      ibex::Array<ibex::Ctc> list_of_contractors;
      for(auto &it : ctc_union->list)
        list_of_contractors.add(*get_complement_contractor(&it));
      return new ibex::CtcCompo(list_of_contractors);
} // etc

      std::ostringstream oss;
      oss << "Type: " << type;
      log_debug("{}", oss.str());

     return nullptr;
```

Note that you don't need `complement_contractor` at all here.